### PR TITLE
note (v5)

### PR DESCRIPTION
**Pull request by rvilarl**
This PR updates `Note` and its main derivates: `StaveNote`, `StemmableNote`, `StaveNote`, `GraceNote`, `NoteHead`

Some functions in `Tables` have been duplicated with and `...Old` variant to avoid having to change the complete library in this PR.

### CODE REVIEW

**Comment by rvilarl** [comment]
Some examples (current first)

pptr-Accidental Basic Bravura svg_current
pptr-Accidental Basic Bravura svg_reference
pptr-canvas_Dot Basic Bravura_current
pptr-canvas_Dot Basic Bravura_reference
pptr-canvas_NoteHead Basic Bravura_current
pptr-canvas_NoteHead Basic Bravura_reference

**Comment by rvilarl** [comment]
Slashes are very different using SMuFL directly.

pptr-Rhythm Rhythm_Draw___slash_notes Bravura svg_current
pptr-Rhythm Rhythm_Draw___slash_notes Bravura svg_reference

**Comment by AaronDavidNewman** [comment]
some flags be kinda whack, too. Looks like maybe some logic with stem height is missing.

image

**Comment by rvilarl** [comment]
> some flags be kinda whack, too. Looks like maybe some logic with stem height is missing.
> 
> image

Thanks Aaron, I will look into that. I tried to adjust it to BB recommendations.

**Comment by rvilarl** [comment]
@AaronDavidNewman stem height fixed, thanks

**Comment by AaronDavidNewman** [comment]
In the formatter test cases, I see this discrepancy.  Rest in different voice overlaps the note.

image

**Comment by AaronDavidNewman** [comment]
The flags with css applied don't 'glow' anymore.  Although the other parts of the note do (in later test cases).  

And the flag doesn't quite meet up with the stem for Gonville font.  Although it was not perfect in the original, so this is pretty subtle.

image

image

**Comment by rvilarl** [comment]
> In the formatter test cases, I see this discrepancy. Rest in different voice overlaps the note.
> 
> image

@AaronDavidNewman now fixed

File: src/notehead.ts [hunk 1]
```diff
@@ -38,66 +29,11 @@ export interface NoteHeadStruct extends NoteStruct {
...
-
-  ctx.restore();
-}
-
 /**
  * `NoteHeads` are typically not manipulated
  * directly, but used internally in `StaveNote`.
  *
- * See `tests/notehead_tests.ts` for usage examples.
+ * See `tests/noteheadTests.ts` for usage examples.
```
Comment (ronyeh) [inline src/notehead.ts:36]: This is a typo (due to camelCase conversion).

Comment (rvilarl) [reply src/notehead.ts]: done

**Review by ronyeh — COMMENTED** [review]
Here are some preliminary comments that you can review.

I will wait until you rebase to do a more in depth review. (And of course, if there is anything that can be extracted into a smaller PR, it will help me greatly in understanding/reviewing.)

File: src/dot.ts [hunk 2]
```diff
@@ -27,16 +27,13 @@ export class Dot extends Modifier {
       if (options?.all) {
         for (let i = 0; i < note.keys.length; i++) {
           const dot = new Dot();
-          dot.setDotShiftY(note.glyphProps.dotShiftY);
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/dot.ts:30]: This was used to shift the Y for rests. Notes take care of that now.

File: src/note.ts [hunk 3]
```diff
@@ -16,17 +14,21 @@ import { Category } from './typeguard';
...
+
 export interface KeyProps {
-  stemDownXOffset?: number;
-  stemUpXOffset?: number;
   key: string;
   octave: number;
   line: number;
   intValue?: number;
-  accidental?: string;
+  acc?: number;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:30]: reverted

File: src/note.ts [hunk 4]
```diff
@@ -92,6 +94,25 @@ export abstract class Note extends Tickable {
...
+    // Lookup duration for default glyph head code
+    let code = Tables.durationCodes[duration];
+    if (code === undefined) {
+      code = Tables.durationCodes['4'];
+    }
+
+    // Try and get the note head
+    const codeNoteHead = Tables.codeNoteHead(type.toUpperCase(), duration);
+    // Merge duration props for 'duration' with the note head properties.
+    if (codeNoteHead != '\u0000') code = { ...code, codeHead: codeNoteHead };
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:111]: done

File: src/note.ts [hunk 5]
```diff
@@ -580,7 +593,7 @@ export abstract class Note extends Tickable {
     let px = this.getRightDisplacedHeadPx();
 
     if (this.checkModifierContext().getMembers('Parenthesis').length !== 0)
-      px += Tables.currentMusicFont().lookupMetric('parenthesis.default.width');
+      px += this.checkModifierContext().getMembers('Parenthesis')[0].getWidth() + 1;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:596]: Takes into account the width of the parenthesis for noteheads within them. Previously it was a constant, now the actual width.

File: src/note.ts [hunk 6]
```diff
@@ -639,7 +652,7 @@ export abstract class Note extends Tickable {
   /** Get point for notes. */
   static getPoint(size?: string): number {
     // for sizes other than 'default', note is 2/3 of the default value
-    return size == 'default' ? Tables.NOTATION_FONT_SCALE : (Tables.NOTATION_FONT_SCALE / 5) * 3;
+    return size == 'default' ? Tables.lookupMetric('fontSize') : (Tables.lookupMetric('fontSize') / 5) * 3;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:655]: done

File: src/notehead.ts [hunk 7]
```diff
@@ -107,18 +43,19 @@ export class NoteHead extends Note {
...
-  protected stemUpXOffset: number = 0;
-  protected stemDownXOffset: number = 0;
   protected displaced: boolean;
   protected stemDirection: number;
 
   protected line: number;
   protected index?: number;
   protected slashed: boolean;
 
+  protected ledger: Record<string, string> = {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/notehead.ts:54]: map to correspoding glyph with ledger line, done

File: src/tabnote.ts [hunk 8]
```diff
@@ -160,10 +160,10 @@ export class TabNote extends StemmableNote {
       // normal glyph scale
       scale: 1.0,
       // default tablature font
-      font: `${Font.SIZE}pt ${Font.SANS_SERIF}`,
+      font: Tables.lookupMetric('fontFamily'),
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/tabnote.ts:163]: yes in CommonMetrics

File: src/tabnote.ts [hunk 9]
```diff
@@ -213,38 +213,41 @@ export class TabNote extends StemmableNote {
...
+
+  static tabToElement(fret: string, scale: number = 1.0): Element {
+    let el: Element;
+
+    if (fret.toUpperCase() === 'X') {
+      el = new Element('TabNote');
+      el.setText(String.fromCodePoint(0xe263 /*accidentalDoubleSharp*/));
+      el.measureText();
+    } else {
+      el = new Element('TabNote.text');
```
Comment (rvilarl) [inline src/tabnote.ts:231]: Here is the font size.

File: tests/tabnote_tests.ts [hunk 10]
```diff
@@ -384,7 +370,7 @@ function drawStemsUpThrough(options: TestOptions, contextBuilder: ContextBuilder
     return tabNote;
   });
 
-  ctx.setFont(Font.SANS_SERIF, 10, FontWeight.BOLD);
+  ctx.setFont(Tables.lookupMetric('fontFamily'), 10, 'bold');
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply tests/tabnote_tests.ts:373]: I suggest ti do that on a dedicated PR

File: src/stavenote.ts [hunk 11]
```diff
@@ -830,9 +780,31 @@ export class StaveNote extends StemmableNote {
...
+        break;
+      case '\ue4e8' /*rest32nd*/:
+      case '\ue4e9' /*rest64th*/:
+        restShift -= 1.5;
+        break;
+      case '\ue4ea' /*rest128th*/:
+        restShift -= 2.5;
+        break;
+    }
+
```
Comment (rvilarl) [inline src/stavenote.ts:804]: This is the yShift that ssuperseeds those in got.ts.

**Comment by ronyeh** [comment]
Let me know once this is rebased. I can probably review this one directly.

**Comment by rvilarl** [comment]
@ronyeh rebased :)

**Comment by rvilarl** [comment]
Nice to see that MuseJazz is now showin the notes correctly!!! Grace notes are also now in 2/3 ratio

current
Grace_Notes Grace_Note_Stem MuseJazz svg_current
reference
Grace_Notes Grace_Note_Stem MuseJazz svg_reference

File: src/tabnote.ts [hunk 12]
```diff
@@ -213,38 +213,41 @@ export class TabNote extends StemmableNote {
...
-        : glyphProps.tabnoteStemDownExtension;
+    return this.flag.getHeight() > Stem.HEIGHT ? this.flag.getHeight() - Stem.HEIGHT : 0;
+  }
+
+  static tabToElement(fret: string, scale: number = 1.0): Element {
+    let el: Element;
+
+    if (fret.toUpperCase() === 'X') {
+      el = new Element('TabNote');
+      el.setText(String.fromCodePoint(0xe263 /*accidentalDoubleSharp*/));
```
Comment (ronyeh) [inline src/tabnote.ts:228]: Do we still need to use String.fromCodePoint()? I thought you were just using the unicode strings directly.

File: src/tabnote.ts [hunk 13]
```diff
@@ -130,7 +130,7 @@ export class TabNote extends StemmableNote {
   }
 
   protected ghost: boolean;
-  protected glyphPropsArr: GlyphProps[] = [];
+  protected fretEl: Element[] = [];
```
Comment (ronyeh) [inline src/tabnote.ts:133]: I think this should either be `fret` or `fretElement`. I don't think the name `fretEl` is consistent with the other times you've included an Element object as an instance var.

File: src/tabnote.ts (hunk 9 above)
Comment (ronyeh) [reply src/tabnote.ts:231]: ? I don't see where the font size is specified here.

File: src/tabnote.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ronyeh) [reply src/tabnote.ts:163]: OK, I'll believe you, but it looks weird that before, the `font` render parameter was specified in CSS form (`10pt Arial`) and now it is just the fontFamily. So this block no longer specifies the font size.

If it's all in CommonMetrics, why do we even specify the font family here? Should we just remove it, and let the CommonMetrics handle it?

**Review by ronyeh — CHANGES_REQUESTED** [review]

File: src/note.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:598]: done

File: src/tabnote.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/tabnote.ts:163]: The render options `font`, `glyphFontScale` and probably `scale` make no longer much sense. The scale concept in general should be reworked. This should be a different PR.

File: src/tabnote.ts (hunk 9 above)
Comment (rvilarl) [reply src/tabnote.ts:231]: In tables:
```
  TabNote: {
    text: {
      fontSize: 9,
    },
  },
```

File: src/accidental.ts [hunk 14]
```diff
@@ -174,14 +174,14 @@ export class Accidental extends Modifier {
       // if this accidental is not a flat, the accidental needs 3.0 lines lower
       // clearance instead of 2.5 lines for b or bb.
       if (
-        accidentalLinePositionAndSpaceNeeds.acc.type !== 'b' &&
-        accidentalLinePositionAndSpaceNeeds.acc.type !== 'bb'
+        accidentalLinePositionAndSpaceNeeds.accidental.type !== 'b' &&
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/accidental.ts:177]: done

File: src/note.ts [hunk 15]
```diff
@@ -16,7 +14,7 @@ import { Category } from './typeguard';
 import { defined, RuntimeError } from './util';
 import { Voice } from './voice';
 
-export interface GlyphPropsNew {
+export interface GlyphProps {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:17]: glyph.ts will be deleted as part of the migration.

File: src/note.ts [hunk 16]
```diff
@@ -665,8 +653,9 @@ export abstract class Note extends Tickable {
 
   /** Get point for notes. */
   static getPoint(size?: string): number {
+    const fontSize = Tables.lookupMetric('fontSize');
     // for sizes other than 'default', note is 2/3 of the default value
-    return size == 'default' ? Tables.NOTATION_FONT_SCALE : (Tables.NOTATION_FONT_SCALE / 5) * 3;
+    return size === 'default' ? fontSize : (fontSize / 5) * 3;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:658]: done

File: src/stemmablenote.ts [hunk 17]
```diff
@@ -253,7 +231,7 @@ export abstract class StemmableNote extends Note {
   }
 
   hasFlag(): boolean {
-    return Tables.getGlyphProps(this.duration).flag == true && !this.beam;
+    return this.glyphProps.codeFlagUp != undefined && !this.beam;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stemmablenote.ts:234]: reworked

File: src/tabnote.ts (hunk 13 above)
Comment (rvilarl) [reply src/tabnote.ts]: done

File: src/stavenote.ts [hunk 18]
```diff
@@ -1108,9 +1080,9 @@ export class StaveNote extends StemmableNote {
 
   shouldDrawFlag(): boolean {
     const hasStem = this.stem !== undefined;
-    const hasFlag = this.glyphProps.flag == true;
+    const hasFlag = this.glyphProps.codeFlagUp != undefined;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stavenote.ts]: done

File: src/tabnote.ts (hunk 12 above)
Comment (rvilarl) [reply src/tabnote.ts]: done

File: src/stemmablenote.ts [hunk 19]
```diff
@@ -56,15 +55,14 @@ export abstract class StemmableNote extends Note {
   buildFlag(): void {
     const { glyphProps } = this;
 
-    if (this.hasFlag()) {
+    if (glyphProps.codeFlagUp && !this.beam && !this.isRest()) {
       const flagCode =
-        this.getStemDirection() === Stem.DOWN ? glyphProps.codeFlagDownstem ?? '' : glyphProps.codeFlagUpstem ?? '';
+        this.getStemDirection() === Stem.DOWN
+          ? String.fromCodePoint((glyphProps.codeFlagUp.codePointAt(0) ?? -1) + 1)
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stemmablenote.ts]: done

File: src/note.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (ronyeh) [reply src/note.ts:17]: Interesting. Are there any APIs in the glyph class that developers (like Aaron) need? If there are useful methods in there, maybe we can keep them around (or reintroduce them after the migration). 

I'm OK with deleting it as part of the migration if it no longer has any obvious uses. But if developers complain, we can look at our v4 branch to see what can be brought back.

File: src/tabnote.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ronyeh) [reply src/tabnote.ts:163]: Yes, if the render options are just hard coded in the constructor, they might as well be moved over to the CommonMetrics. If the constructor accepts optional render options, they can then override the value we looked up from the CommonMetrics.

File: src/stemmablenote.ts [hunk 20]
```diff
@@ -253,7 +232,7 @@ export abstract class StemmableNote extends Note {
   }
 
   hasFlag(): boolean {
-    return Tables.getGlyphProps(this.duration).flag == true && !this.beam;
+    return this.glyphProps.codeFlagUp != undefined && !this.beam && !this.isRest();
```
Comment (ronyeh) [inline src/stemmablenote.ts:235]: @rvilarl I'm going to go ahead and merge this, but in the next PR can you change this != to !== unless there is a great reason to use the looser inequality check?

File: src/stemmablenote.ts [hunk 21]
```diff
@@ -58,13 +57,13 @@ export abstract class StemmableNote extends Note {
 
     if (this.hasFlag()) {
       const flagCode =
-        this.getStemDirection() === Stem.DOWN ? glyphProps.codeFlagDownstem ?? '' : glyphProps.codeFlagUpstem ?? '';
+        // codeFlagDown = codeFlagUp + 1,, if not defined, code should be 0
+        this.getStemDirection() === Stem.DOWN
+          ? String.fromCodePoint((glyphProps.codeFlagUp?.codePointAt(0) ?? -1) + 1)
+          : glyphProps.codeFlagUp ?? '\u0000';
```
Comment (ronyeh) [inline src/stemmablenote.ts:63]: @rvilarl I'm going to merge, but this block could still be improved, I think. :-)

File: src/stavenote.ts [hunk 22]
```diff
@@ -692,7 +641,8 @@ export class StaveNote extends StemmableNote {
    * @returns true if this note is a type of rest. Rests don't have pitches, but take up space in the score.
    */
   isRest(): boolean {
-    return this.glyphProps.rest;
+    const val = this.glyphProps.codeHead;
+    return val >= '\ue4e0' && val <= '\ue4ff';
```
Comment (ronyeh) [inline src/stavenote.ts:645]: @rvilarl Does this work? What type is codeHead? Is it a string? Can we do inequality checks with the strings here?

I think it is safer to do this type of check with numeric values.

File: src/stavenote.ts [hunk 23]
```diff
@@ -427,7 +408,7 @@ export class StaveNote extends StemmableNote {
     this.renderOptions = {
       ...this.renderOptions,
       // font size for note heads and rests
-      glyphFontScale: noteStruct.glyphFontScale || Tables.NOTATION_FONT_SCALE,
+      glyphFontScale: noteStruct.glyphFontScale || Tables.lookupMetric('fontSize'),
```
Comment (ronyeh) [inline src/stavenote.ts:411]: If the first item can be undefined, we should use ?? instead.

**Review by ronyeh — CHANGES_REQUESTED** [review]
I'm going to go ahead and merge, but please see my most recent comments for suggested changes that you can include in the next PR. Thanks!

File: src/stavenote.ts (hunk 22 above)
Comment (ronyeh) [reply src/stavenote.ts:645]: From my most recent review, I think this one is the most confusing. We shouldn't really be doing mathematical comparisons against string values.

**Pull request merged**